go-controller/pkg/node/gateway_shared_intf.go

```
@@ -1581,23 +1586,121 @@ func svcToCookie(namespace string, name string, token string, port int32) (strin
+			_, masqIPNet, _ = net.ParseCIDR(fmt.Sprintf("%s/128", types.V6OVNMasqueradeIP))
+			v6Found = true
+		} else if !v4Found {
+			_, masqIPNet, _ = net.ParseCIDR(fmt.Sprintf("%s/32", types.V4OVNMasqueradeIP))
+			v4Found = true
+		}
+		mtu := config.Default.MTU
+		if config.Default.RoutableMTU != 0 {
+			mtu = config.Default.RoutableMTU
+		}
+		if masqIPNet != nil {
+			klog.Infof("Setting OVN Masquerade route with source: %s", nodeIP)
+			err = util.LinkRoutesAddOrUpdateSourceOrMTU(netIfaceLink, nil, []*net.IPNet{masqIPNet},
+				mtu, nodeIP)
+			if err != nil {
+				return fmt.Errorf("unable to add OVN masquerade route to host, error: %v", err)
+			}
+		}
+	}
+
+	if config.IPv4Mode && !v4Found {
+		return fmt.Errorf("could not find node IPv4 address to configure OVN masquerade route, addresses: %+v",
+			node.Status.Addresses)
+	}
+
+	if config.IPv6Mode && !v6Found {
+		return fmt.Errorf("could not find node IPv6 address to configure OVN masquerade route, addresses: %+v",
+			node.Status.Addresses)
+	}
+	return nil
+}
+
+func setNodeMasqueradeIPOnExtBridge(extBridgeName string) error {
+	extBridge, err := util.LinkSetUp(extBridgeName)
 	if err != nil {
-		return fmt.Errorf("no valid ipv4 next hop exists: %v", err)
+		return err
+	}
+
+	var bridgeCIDRs []*net.IPNet
+	if config.IPv4Mode {
+		_, masqIPNet, _ := net.ParseCIDR(types.V4MasqueradeSubnet)
+		masqIPNet.IP = net.ParseIP(types.V4HostMasqueradeIP)
+		bridgeCIDRs = append(bridgeCIDRs, masqIPNet)
 	}
-	_, masqIPNet, _ := net.ParseCIDR(types.V4MasqueradeSubnet)
-	err = util.LinkRoutesAddOrUpdateMTU(netIfaceLink, v4nextHops[0], []*net.IPNet{masqIPNet}, config.Default.RoutableMTU)
+
+	if config.IPv6Mode {
+		_, masqIPNet, _ := net.ParseCIDR(types.V6MasqueradeSubnet)
+		masqIPNet.IP = net.ParseIP(types.V6HostMasqueradeIP)
+		bridgeCIDRs = append(bridgeCIDRs, masqIPNet)
+	}
+
+	for _, bridgeCIDR := range bridgeCIDRs {
+		if exists, err := util.LinkAddrExist(extBridge, bridgeCIDR); err == nil && !exists {
```

hmm we probably should not ignore this err here